examples/erc20/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+
+    /// Account holders
+    accounts: HashMap<Address, u64>,
+
+    /// Allowances
+    allowed: HashMap<Address, HashMap<Address, u64>>,
+}
+
+// A Transfer event struct
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Event)]
+pub struct Transfer {
+    #[indexed]
+    pub from: Address,
+    #[indexed]
+    pub to: Address,
+    #[indexed]
+    pub amount: u64,
+}
+
+// An Approval event struct
+#[derive(Serialize, Deserialize, Clone, Debug, Default, Event)]
+pub struct Approval {
+    #[indexed]
+    pub sender: Address,
+    #[indexed]
+    pub spender: Address,
+    #[indexed]
+    pub amount: u64,
+}
+
+impl ERC20Token {
+    /// Constructs a new `ERC20Token`
+    pub fn new(ctx: &Context, total_supply: u64) -> Result<Self> {
+        let owner = ctx.sender();
+        let mut admins = HashSet::new();
+        admins.insert(owner);
+        let mut accounts = HashMap::new();
+        accounts.insert(owner, total_supply);
+
+        Ok(Self {
+            total_supply,
+            owner,
+            admins,
+            accounts,
+            ..Default::default()
+        })
+    }
+
+    /// Get balance
+    pub fn balance_of(&mut self, ctx: &Context) -> Result<u64> {
+        let balance = match self.accounts.entry(ctx.sender()) {
+            Entry::Vacant(_ve) => 0u64,
+            Entry::Occupied(oe) => *oe.get(),
+        };
+        Ok(balance)
+    }
+
+    /// Get total supply
+    pub fn total_supply(&mut self, _ctx: &Context) -> Result<u64> {
+        Ok(self.total_supply)
+    }
+
+    /// Add admin
+    pub fn add_admin(&mut self, ctx: &Context, admin: Address) -> Result<()> {
+        if !self.admins.contains(&ctx.sender()) {
+            return Err(Error::AdminPrivilegesRequired);
+        }
+        self.admins.insert(admin);
+        Ok(())
+    }
+}
+
+// Helper methods
+
+/// Balance of
+fn _balance(accounts: &mut HashMap<Address, u64>, addr: Address) -> u64 {
+    match accounts.entry(addr) {
+        Entry::Vacant(_ve) => return 0u64,
+        Entry::Occupied(oe) => return *oe.get(),
+    }
+}
+
+/// transfer method
+fn _transfer(
+    accounts: &mut HashMap<Address, u64>,
+    from: Address,
+    to: Address,
+    amount: u64,
+) -> bool {
+    let from_balance = _balance(accounts, from);
+    let to_balance = _balance(accounts, to);
+
+    // check for sufficient balance
+    if from_balance < amount {
+        return false;
+    }
+    accounts.insert(from, from_balance - amount);
+    accounts.insert(to, to_balance + amount);
+
+    Event::emit(&Transfer {
+        from,
+        to,
+        amount,
+    });
+
+    true
+}
+
+impl ERC20Token {
+    /// transfer
+    pub fn transfer(&mut self, ctx: &Context, to: Address, amount: u64) -> Result<Transfer> {
+        let from = ctx.sender();
+        if from == to || amount == 0u64 {
+            // no-op
+            return Ok(Transfer::default());
+        }
+        if _transfer(&mut self.accounts, ctx.sender(), to, amount) {
```

Comment:
why not have `_transfer` return `Result` directly?

Reply:
Using it for `transfer_from` and `transfer`. The result is  used differently in those two fellows.